### DISCUSSION TITLE
Help with spurious clangd errors with C++20 mode?

### DISCUSSION THREAD

**Discussion by ahrbe1** [clangd/clangd#1847] [Q&A] [answered]
I'm seeing an issue with my primary Windows 10 development machine where clangd-based static analysis generates lots of bogus errors when in C++20 mode. Even the basic "hello world" program generates some weird errors.

Sample code:

```cmake
cmake_minimum_required(VERSION 3.26)  
project(broken)  

set(CMAKE_CXX_STANDARD 20)  
set(CMAKE_CXX_STANDARD_REQUIRED ON)  
set(CMAKE_EXPORT_COMPILE_COMMANDS ON)  

add_executable(broken main.cpp)
```

```c++
#include <iostream>  
int main() {  
 std::cout << "Hello, World!" << std::endl;  
 return 0;  
}
```

<img width="821" alt="Screenshot 2023-12-03 140136" src="https://github.com/clangd/clangd/assets/6035787/b952d4eb-d4a9-4224-bacd-bfe79cdcb010">

The error occurs in the `<<` operator (see screenshot):

```
In template: constexpr variable 'is_integral_v<const char *>' must be initialized by a constant expression
```

which occurs in this snippet of code from a header called `xtr1common`:

```c++
_EXPORT_STD template <class _Ty>
_INLINE_VAR constexpr bool is_integral_v = _Is_any_of_v<remove_cv_t<_Ty>, bool, char, signed char, unsigned char,
    wchar_t,
#ifdef __cpp_char8_t
    char8_t,
#endif // defined(__cpp_char8_t)
    char16_t, char32_t, short, unsigned short, int, unsigned int, long, unsigned long, long long, unsigned long long>;
```

This happens with both CLion's embedded clangd as well as VSCode's clangd plugin. I've uninstalled and reinstalled Visual Studio 2019 and 2022, and tried repairing the installations, upgrading to preview versions, etc. But neither CLion nor VSCode's clangd analyser works for me in C++20 mode. I've spent probably 3 days so far trying to debug this. No luck yet, hence my post here. 

C++17 mode works as expected, though. 

From inspecting `xtr1common`, it seems that if I compile with `-DHAVE_STD_BYTE=0` then this particular error goes away. But I usually bump into other errors with `is_integral_v`, for example, when using `std::make_shared`, or creating a `std::shared_ptr`.

FWIW, clangd versions is 17.0.3 (VSCode) and 17.0.0 (CLion)

Has anybody seen any similar issues with clangd, or have ideas on how to fix this? This machine is basically unusable with C++20 due to this issue, and I am struggling to find a fix.

**Comment by HighCommander4** [comment]
Can you **build** your code in C++20 mode with clang as the compiler?

**Comment by ahrbe1** [comment]
Yes, I just tried compiling it with llvm/clang 17.0.6 and that works fine, however the clangd errors did not go away when I switched the compiler in the IDE. I ran the build in verbose mode and confirmed that it is indeed building sucessfully with clang.

**Reply by HighCommander4** [reply]
Ok, interesting. Could you share [verbose clangd logs](https://clangd.llvm.org/troubleshooting#gathering-logs) from VSCode please?

**Comment by ahrbe1** [comment, marked as answer]
Well, so in trying to gather the logs for this, I added this to my project's `settings.json`:

```json
    "clangd.arguments": [
        "--log=verbose"
    ],
```

In looking through the log files before uploading them, I noticed that somehow there was a `-std=c++17` flag in there which shouldn't be in there. Well, it took me basically all evening to figure out where that was coming from... eventually I found a config file at `AppData\Local\clangd\config.yaml` which had that flag in it. So I deleted the config file and the index and forced everything to re-index, and now both CLion and VSCode seem to be working properly. I have no idea where that config file came from, but removing it seemed to have resolved this issue. 

Thanks!
